mne/time_frequency/tfr.py

```diff
@@ -546,6 +547,139 @@ def _time_frequency_loop(X, Ws, output, use_fft, mode, decim):
+            if method == 'morlet':
+                # make a list to have the same dims as the 'multitaper' case
+                Ws = [morlet(sfreq, freqs, n_cycles=n_cycles,
+                             zero_mean=zero_mean)]
+
+            elif method == 'multitaper':
+                Ws = _make_dpss(sfreq, freqs, n_cycles=n_cycles,
+                                time_bandwidth=time_bandwidth,
+                                zero_mean=zero_mean)
+
+            # check tapers
+            if len(Ws[0][0]) > inst.shape[-1]:
+                raise ValueError('At least one of the wavelets is longer '
+                                 'than the signal. Use a longer signal or '
+                                 'shorter wavelets.')
+
+            # initiate tfrs
+            n_times = X.shape[-1]
+            n_verts = X.shape[0]
+            if ('avg_' in output) or ('itc' in output):
+                tfrs = np.zeros((n_verts, n_freqs, n_times), dtype=dtype)
+            else:
+                tfrs = np.zeros((1, n_verts, n_freqs, n_times), dtype=dtype)
+
+            # Inter-trial phase locking is summed up for each taper
+            if 'itc' in output:
+                plf = np.zeros((len(Ws), n_verts, n_freqs, n_times),
+                               dtype=np.complex)
+
+        else:  # for all iterations except the first
+
+            # add a new epoch element to the tfrs array after each epoch
+            if not (('avg_' in output) or ('itc' in output)):
+                new_epoch = np.zeros((1, n_verts, n_freqs, n_times),
+                                     dtype=dtype)
+                tfrs = np.concatenate((tfrs, new_epoch), axis=0)
+
+            # make sure these elements got the same properties as the first one
+            if not isinstance(inst, type_ref):
+                raise TypeError("All computed elements must be of the same "
+                                "SourceEstimate type. Got {} and {}."
+                                .format(type_ref, type(inst)))
+            if not inst.sfreq == sfreq:
+                raise ValueError("All computed elements must have the same "
+                                 "sfreq.")
+            if not inst.tmin == tmin_ref:
+                raise ValueError("All computed elements must have the same "
+                                 "tmin.")
+
+        # loop over the tapers
+        for W_idx, W in enumerate(Ws):
+
+            tfr = cwt(X, W, use_fft=use_fft, mode='same', decim=decim)
+
+            # Transform complex values
+            if output in ['power', 'avg_power']:
+                tfr = (tfr * tfr.conj()).real  # power
+            elif output == 'phase':
+                tfr = np.angle(tfr)
+            elif output == 'avg_power_itc':
+                tfr_abs = np.abs(tfr)
+                plf[W_idx] += tfr / tfr_abs  # phase
+                tfr = tfr_abs ** 2  # power
+            elif output == 'itc':
+                plf[W_idx] += tfr / np.abs(tfr)  # phase
```

Comment:
@agramfort This here is the critical part we will have to handle, if we want to allow tfr_functions to work economically with (kernel, sens_data) type data.
As you can see there are a lot of non-linear operations in here. If we want to keep the kernel until the creation of the SourceTFR, we will have to shift all these steps into e.g. the `stfr._remove_kernel_sens_data_` method. This itself wouldn't be the problem and was already implemented for non-averaged power. However, if we want to average, or get the ITC, we encounter the new problem that these steps are performed after the non-linear operation, and we would also need to carry them over to SourceTFR. For average, that would mean that we had to additionally pass the epochs axis to the SourceTFR, and then, while computing the data field, first do `np.abs(data) ** 2` and later average it. For ITC data, I think we would need to even go a step further and add a further axis to the sensor data, representing the tapers (since the ITCs need to be processed along the taper axis, (See line 669).
So, this solution can be done, but it will look really nasty and counter-intuitive for anyone who will have to handle the code. On the other hand, this will allow to save memory up to the point where the user calls the data field of the SourceTFR.

A second, quite simple solution (for which I already made a test implementation) would be to retain the kernel until line 644, where the continuous wavelet transform is processed, and construct the full data right after that. This means, we can compute the `cwt` on our slim set of sensor space channels, but do all of the following computations in Source Space.
Since the full data will need to be computed anyways at some point, and we will also need to perform the non-linear steps (+ following steps) after computing the full data, I think this will result in an equal memory consumption.

What do you think about this?

Reply:
The slow part is usually the wavelet/fft step, so it seems reasonable to transform after that. This is especially true if the alternative leads to more complicated code, which it sounds like it will.

We can always improve it later if we think of a way.

Reply:
Yes, I think that's the best option for now!

Reply:
As for the timing, the second implementation looks quite good. I made some simple timing runs on my laptop using timeit.default_timer. I compared tfr_morlet with delayed and undelayed data, as well as source_induced_power with the same parameters and data.

for an stc of 7928 verts x 3 oris x 20 eps x 211 times, average runtimes were:
* 52.050 seconds for source_induced_power
* 43.922 seconds for tfr_morlet with delayed data
* 75.948 seconds for tfr_morlet with undelayed data

for an stc of 33 verts x 3 oris x 20 eps x 211 times, average runtimes were:
* 1.621 seconds for source_induced_power
* 1.029 seconds for tfr_morlet with delayed data
* 0.296 seconds for tfr_morlet with undelayed data (This is shorter now because we've got less vertices than channels)